File: link-checker.py
#!/usr/bin/env python3
"""
Manual link checker for nextellar-docs
Replicates the functionality of scripts/check-links.cjs
"""

import os
import re
from pathlib import Path
from collections import defaultdict

# Colors for output
class Colors:
    RESET = '\033[0m'
    RED = '\033[31m'
    GREEN = '\033[32m'
    YELLOW = '\033[33m'
    BLUE = '\033[34m'
    CYAN = '\033[36m'

DOCS_DIR = Path(__file__).parent / 'docs'
DOCS_BASE = '/docs'

def walk_docs(directory=DOCS_DIR):
    """Recursively collect all .mdx and .md files"""
    file_list = []
    for root, dirs, files in os.walk(directory):
        for file in files:
            if file.endswith(('.mdx', '.md')):
                file_list.append(Path(root) / file)
    return sorted(file_list)

def file_path_to_route(file_path):
    """Convert file path to URL route"""
    relative = file_path.relative_to(DOCS_DIR)
    without_ext = re.sub(r'\.(mdx|md)$', '', str(relative))
    normalized = without_ext.replace('\\', '/')
    
    # Handle index files
    if normalized.endswith('/index'):
        return DOCS_BASE + '/' + normalized.replace('/index', '')
    
    return DOCS_BASE + '/' + normalized

def extract_anchors(content):
    """Extract all headings and id attributes from markdown/html"""
    anchors = set()
    
    # Markdown headings: # Heading, ## Subheading, etc.
    heading_pattern = r'^#+\s+(.+?)(?:\s*\{.*?\})?$'
    for match in re.finditer(heading_pattern, content, re.MULTILINE):
        heading = match.group(1).strip()
        # Convert heading to slug
        slug = heading.lower()
        slug = re.sub(r'[^\w\s-]', '', slug)
        slug = slug.strip()
        slug = re.sub(r'\s+', '-', slug)
        slug = re.sub(r'-+', '-', slug)
        if slug:
            anchors.add(slug)
    
    # HTML id attributes
    id_pattern = r"id=['\"]([\\w-]+)['\"]"
    for match in re.finditer(id_pattern, content):
        anchors.add(match.group(1))
    
    return anchors

def extract_links(content):
    """Extract internal links from markdown"""
    links = []
    lines = content.split('\n')
    
    for line_num, line in enumerate(lines, 1):
        # Markdown links: [text](url)
        md_link_pattern = r'\[([^\]]+)\]\(([^)]+)\)'
        for match in re.finditer(md_link_pattern, line):
            href = match.group(2)
            # Only internal links
            if not href.startswith(('http', 'mailto:', 'tel:')):
                links.append({
                    'href': href,
                    'type': 'markdown',
                    'lineNum': line_num,
                    'text': match.group(1)
                })
        
        # HTML href attributes
        html_href_pattern = r"href=['\"]([^'\"]+)['\"]"
        for match in re.finditer(html_href_pattern, line):
            href = match.group(1)
            # Only internal links
            if not href.startswith(('http', 'mailto:', 'tel:')):
                links.append({
                    'href': href,
                    'type': 'html',
                    'lineNum': line_num
                })
    
    return links

def normalize_href(href):
    """Normalize link href for comparison"""
    # Remove trailing slashes and query params/fragments
    normalized = href.split('#')[0].split('?')[0].rstrip('/')
    
    if not normalized.startswith('/'):
        return normalized
    
    # Ensure /docs prefix
    if not normalized.startswith('/docs/'):
        normalized = '/docs' + normalized
    
    return normalized

def resolve_link(href, from_route, known_routes, anchors):
    """Check if a link target exists"""
    # Anchor-only links
    if href.startswith('#'):
        anchor = href[1:]
        page_anchors = anchors.get(from_route, set())
        if anchor in page_anchors:
            return {'valid': True, 'reason': 'anchor exists'}
        return {'valid': False, 'reason': f"anchor '#{anchor}' not found on this page"}
    
    # Normalize the href
    target_route = normalize_href(href)
    
    # Check if route exists
    if target_route in known_routes:
        return {'valid': True, 'reason': 'page exists'}
    
    # Check without /docs variant
    without_docs = target_route.replace('/docs', '', 1)
    if without_docs != target_route and without_docs in known_routes:
        return {'valid': True, 'reason': 'page exists'}
    
    return {'valid': False, 'reason': 'page not found'}

def main():
    print(f"{Colors.CYAN}🔗 STATIC LINK CRAWLER{Colors.RESET}\n")
    print(f"Scanning docs directory: {DOCS_DIR}\n")
    
    # Collect all doc files
    doc_files = walk_docs()
    print(f"{Colors.BLUE}Found {len(doc_files)} documentation files{Colors.RESET}\n")
    
    # Build map of known routes and anchors
    known_routes = set()
    anchors = {}
    
    for file_path in doc_files:
        route = file_path_to_route(file_path)
        known_routes.add(route)
        
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
            anchors[route] = extract_anchors(content)
    
    # Extract and validate links
    broken_links = []
    total_links = 0
    valid_links = 0
    
    for file_path in doc_files:
        route = file_path_to_route(file_path)
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        links = extract_links(content)
        
        for link in links:
            # Skip anchors and external links
            if link['href'].startswith('#') or link['href'].startswith('http'):
                continue
            
            total_links += 1
            result = resolve_link(link['href'], route, known_routes, anchors)
            
            if not result['valid']:
                broken_links.append({
                    'file': file_path,
                    'route': route,
                    'href': link['href'],
                    'lineNum': link['lineNum'],
                    'type': link['type'],
                    'text': link.get('text', ''),
                    'reason': result['reason']
                })
            else:
                valid_links += 1
    
    # Report results
    print(f"{Colors.BLUE}Validation Results:{Colors.RESET}")
    print(f"✅ Valid links: {valid_links}/{total_links}")
    
    if broken_links:
        print(f"{Colors.RED}❌ Broken links: {len(broken_links)}{Colors.RESET}\n")
        print(f"{Colors.YELLOW}Broken Links Report:{Colors.RESET}\n")
        
        # Group by file
        links_by_file = defaultdict(list)
        for link in broken_links:
            links_by_file[link['route']].append(link)
        
        idx = 1
        for route in sorted(links_by_file.keys()):
            print(f"{Colors.CYAN}{route}{Colors.RESET}")
            for link in links_by_file[route]:
                print(f"  {Colors.RED}[{idx}]{Colors.RESET} Line {link['lineNum']}")
                print(f"       {Colors.YELLOW}Link: {link['href']}{Colors.RESET}")
                print(f"       {Colors.RED}Error: {link['reason']}{Colors.RESET}")
                if link['text']:
                    print(f"       Text: \"{link['text']}\"")
                idx += 1
            print()
        
        print(f"{Colors.RED}❌ VALIDATION FAILED - {len(broken_links)} broken link(s) found{Colors.RESET}")
        return 1
    else:
        print(f"{Colors.GREEN}🎉 ALL LINKS VALID!{Colors.RESET}")
        return 0

if __name__ == '__main__':
    exit(main())

File: scripts/link-checker.mjs
#!/usr/bin/env node

/**
 * Robust recursive link checker for nextellar-docs
 * Validates internal links (with anchor support) and optionally checks external links
 * 
 * Usage:
 *   node scripts/link-checker.mjs [options]
 * 
 * Options:
 *   --check-external    Enable external link checking (default: disabled)
 *   --timeout NUMBER    Request timeout in ms (default: 5000)
 *   --max-concurrent N  Max concurrent external requests (default: 5)
 *   --allowlist FILE    JSON file with URLs to skip
 *   --verbose           Verbose output
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { promisify } from 'util';
import http from 'http';
import https from 'https';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const docsRoot = path.join(__dirname, '..', 'docs');

// ============================================================================
// CONFIGURATION
// ============================================================================

const config = {
  checkExternal: process.argv.includes('--check-external'),
  timeout: parseInt(process.argv.find(arg => arg.startsWith('--timeout='))?.split('=')[1] || '5000'),
  maxConcurrent: parseInt(process.argv.find(arg => arg.startsWith('--max-concurrent='))?.split('=')[1] || '5'),
  allowlistFile: process.argv.find(arg => arg.startsWith('--allowlist='))?.split('=')[1],
  verbose: process.argv.includes('--verbose'),
};

// ============================================================================
// UTILITIES
// ============================================================================

/**
 * Generate heading ID from text using remark-slug logic
 * Matches: https://github.com/remarkjs/remark-slug
 */
function generateHeadingId(text) {
  return text
    .toLowerCase()
    .trim()
    .replace(/\s+/g, '-')          // Replace spaces with hyphens
    .replace(/[^\w-]/g, '')         // Remove non-word chars except hyphens
    .replace(/--+/g, '-')           // Collapse multiple hyphens
    .replace(/^-+|-+$/g, '');       // Remove leading/trailing hyphens
}

/**
 * Extract all markdown headings from content
 * Returns array of { level, text, id }
 */
function extractHeadings(content) {
  const headingRegex = /^(#{1,6})\s+(.+)$/gm;
  const headings = [];
  let match;

  while ((match = headingRegex.exec(content)) !== null) {
    const level = match[1].length;
    const text = match[2].trim();
    const id = generateHeadingId(text);
    headings.push({ level, text, id });
  }

  return headings;
}

/**
 * Extract all links from markdown/MDX content
 * Returns array of { text, href, line, isExternal }
 */
function extractLinks(content) {
  const links = [];
  
  // Markdown links: [text](url)
  const markdownLinkRegex = /\[([^\]]+)\]\(([^)]+)\)/g;
  let match;
  let lineNum = 1;

  const lines = content.split('\n');
  
  lines.forEach((line, idx) => {
    let lineMatch;
    while ((lineMatch = markdownLinkRegex.exec(line)) !== null) {
      const text = lineMatch[1];
      const href = lineMatch[2];
      const isExternal = href.startsWith('http://') || href.startsWith('https://');
      
      links.push({
        text,
        href,
        line: idx + 1,
        isExternal,
      });
    }
  });

  return links;
}

/**
 * Resolve relative link to actual file path
 * Handles: ../file, ./file, /docs/file, docs/file
 */
function resolveLink(fromFile, href) {
  const baseDir = path.dirname(fromFile);
  
  // Remove fragment
  const [linkPath] = href.split('#');
  
  if (!linkPath) return { file: fromFile, fragment: href.substring(1) };

  let resolved;
  
  if (linkPath.startsWith('/')) {
    // Absolute path from docs root
    resolved = path.join(docsRoot, linkPath);
  } else {
    // Relative path
    resolved = path.resolve(baseDir, linkPath);
  }

  // Ensure it's within docs root (security check)
  const normalized = path.normalize(resolved);
  const docsDirNormalized = path.normalize(docsRoot);
  
  if (!normalized.startsWith(docsDirNormalized)) {
    return null; // Outside docs root
  }

  // Add .mdx/.md extension if missing
  let finalPath = normalized;
  if (!finalPath.endsWith('.mdx') && !finalPath.endsWith('.md')) {
    // Try .mdx first, then .md
    if (fs.existsSync(normalized + '.mdx')) {
      finalPath = normalized + '.mdx';
    } else if (fs.existsSync(normalized + '.md')) {
      finalPath = normalized + '.md';
    } else if (fs.existsSync(normalized) && fs.statSync(normalized).isDirectory()) {
      // Check for index.mdx in directory
      const indexPath = path.join(normalized, 'index.mdx');
      if (fs.existsSync(indexPath)) {
        finalPath = indexPath;
      } else {
        finalPath = null;
      }
    } else {
      finalPath = null;
    }
  }

  return {
    file: finalPath,
    fragment: href.includes('#') ? href.split('#')[1] : null,
  };
}

/**
 * Get all MDX/MD files recursively
 */
function getAllDocFiles(dir = docsRoot) {
  const files = [];
  const entries = fs.readdirSync(dir, { withFileTypes: true });

  entries.forEach(entry => {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      files.push(...getAllDocFiles(fullPath));
    } else if (entry.name.endsWith('.mdx') || entry.name.endsWith('.md')) {
      files.push(fullPath);
    }
  });

  return files;
}

/**
 * Load allowlist of URLs to skip
 */
function loadAllowlist() {
  if (!config.allowlistFile) return new Set();
  
  try {
    const content = fs.readFileSync(config.allowlistFile, 'utf-8');
    const data = JSON.parse(content);
    return new Set(data.urls || []);
  } catch (err) {
    console.error(`Failed to load allowlist: ${err.message}`);
    return new Set();
  }
}

/**
 * Format file path relative to docs root for clearer output
 */
function formatPath(fullPath) {
  return path.relative(docsRoot, fullPath);
}

// ============================================================================
// INTERNAL LINK VALIDATION
// ============================================================================

/**
 * Validate internal links in a file
 */
function validateInternalLinks(filePath, content) {
  const issues = [];
  const links = extractLinks(content);
  const internalLinks = links.filter(link => !link.isExternal);

  internalLinks.forEach(link => {
    const resolved = resolveLink(filePath, link.href);

    if (!resolved) {
      issues.push({
        type: 'invalid-path',
        file: formatPath(filePath),
        line: link.line,
        link: link.href,
        text: link.text,
        message: `Link points outside docs root: ${link.href}`,
      });
      return;
    }

    if (!resolved.file) {
      issues.push({
        type: 'broken-link',
        file: formatPath(filePath),
        line: link.line,
        link: link.href,
        text: link.text,
        message: `Target file not found: ${link.href}`,
      });
      return;
    }

    if (resolved.fragment) {
      // Validate anchor
      try {
        const targetContent = fs.readFileSync(resolved.file, 'utf-8');
        const headings = extractHeadings(targetContent);
        const headingIds = headings.map(h => h.id);

        if (!headingIds.includes(resolved.fragment)) {
          issues.push({
            type: 'broken-anchor',
            file: formatPath(filePath),
            line: link.line,
            link: link.href,
            text: link.text,
            targetFile: formatPath(resolved.file),
            fragment: resolved.fragment,
            availableHeadings: headingIds,
            message: `Anchor not found: ${resolved.fragment} in ${formatPath(resolved.file)}`,
          });
        }
      } catch (err) {
        issues.push({
          type: 'anchor-validation-error',
          file: formatPath(filePath),
          line: link.line,
          link: link.href,
          text: link.text,
          message: `Failed to validate anchor: ${err.message}`,
        });
      }
    }
  });

  return issues;
}

// ============================================================================
// EXTERNAL LINK VALIDATION
// ============================================================================

/**
 * Check external link with HEAD request, fallback to GET
 */
async function checkExternalLink(url, attempt = 1) {
  return new Promise((resolve) => {
    const makeRequest = (method) => {
      const urlObj = new URL(url);
      const client = url.startsWith('https') ? https : http;
      const options = {
        hostname: urlObj.hostname,
        port: urlObj.port,
        path: urlObj.pathname + urlObj.search,
        method: method,
        headers: {
          'User-Agent': 'nextellar-link-checker/1.0',
        },
        timeout: config.timeout,
      };

      const req = client.request(options, (res) => {
        resolve({
          url,
          status: res.statusCode,
          method,
          success: res.statusCode < 400,
        });
      });

      req.on('error', (err) => {
        if (method === 'HEAD' && attempt < 2) {
          // Fallback to GET
          makeRequest('GET');
        } else {
          resolve({
            url,
            error: err.message,
            method,
            success: false,
          });
        }
      });

      req.on('timeout', () => {
        req.destroy();
        resolve({
          url,
          error: `Timeout after ${config.timeout}ms`,
          method,
          success: false,
        });
      });

      req.end();
    };

    makeRequest('HEAD');
  });
}

/**
 * Validate external links with concurrency control
 */
async function validateExternalLinks(links) {
  const results = [];
  const queue = links.slice();
  let active = 0;

  return new Promise((resolve) => {
    const processNext = () => {
      if (queue.length === 0 && active === 0) {
        resolve(results);
        return;
      }

      while (active < config.maxConcurrent && queue.length > 0) {
        active++;
        const link = queue.shift();

        checkExternalLink(link.href).then((result) => {
          results.push({
            ...link,
            ...result,
          });
          active--;
          processNext();
        });
      }
    };

    processNext();
  });
}

// ============================================================================
// MAIN CHECKER
// ============================================================================

async function runLinkChecker() {
  console.log('🔗 Nextellar Link Checker\n');

  const allFiles = getAllDocFiles();
  const allowlist = loadAllowlist();
  
  console.log(`📂 Scanning ${allFiles.length} documentation files...\n`);

  let internalIssues = [];
  let externalLinks = [];
  const fileHeadings = new Map();

  // Phase 1: Scan all files for links and headings
  allFiles.forEach(filePath => {
    try {
      const content = fs.readFileSync(filePath, 'utf-8');
      
      // Extract headings for reference
      const headings = extractHeadings(content);
      fileHeadings.set(filePath, headings);

      // Validate internal links
      const issues = validateInternalLinks(filePath, content);
      internalIssues = internalIssues.concat(issues);

      // Collect external links
      if (config.checkExternal) {
        const links = extractLinks(content);
        const externalOnly = links
          .filter(link => link.isExternal && !allowlist.has(link.href))
          .map(link => ({ ...link, sourceFile: filePath }));
        externalLinks = externalLinks.concat(externalOnly);
      }
    } catch (err) {
      console.error(`Error reading file ${formatPath(filePath)}: ${err.message}`);
      process.exit(1);
    }
  });

  // Phase 2: Check external links (if enabled)
  let externalIssues = [];
  if (config.checkExternal && externalLinks.length > 0) {
    console.log(`🌐 Checking ${externalLinks.length} external links...`);
    const uniqueUrls = [...new Set(externalLinks.map(l => l.href))];
    const results = await validateExternalLinks(uniqueUrls.map(href => ({ href })));

    externalIssues = results
      .filter(result => !result.success)
      .flatMap(result => {
        return externalLinks
          .filter(link => link.href === result.url)
          .map(link => ({
            type: 'broken-external-link',
            file: formatPath(link.sourceFile),
            line: link.line,
            link: link.href,
            text: link.text,
            error: result.error || `HTTP ${result.status}`,
            message: `External link failed: ${result.error || `HTTP ${result.status}`}`,
          }));
      });
  }

  // =========================================================================
  // REPORT RESULTS
  // =========================================================================

  console.log('\n' + '='.repeat(70));
  console.log('LINK VALIDATION RESULTS');
  console.log('='.repeat(70) + '\n');

  if (internalIssues.length === 0 && externalIssues.length === 0) {
    console.log('✅ All links valid!\n');
    console.log(`📊 Summary:`);
    console.log(`   Internal links: OK`);
    if (config.checkExternal) {
      console.log(`   External links: OK (${externalLinks.length} checked)`);
    }
    return { success: true, totalIssues: 0 };
  }

  // Group issues by type
  const issuesByType = {};
  const allIssues = [...internalIssues, ...externalIssues];

  allIssues.forEach(issue => {
    if (!issuesByType[issue.type]) {
      issuesByType[issue.type] = [];
    }
    issuesByType[issue.type].push(issue);
  });

  // Report internal issues
  if (internalIssues.length > 0) {
    console.log('❌ INTERNAL LINK ISSUES:\n');

    ['broken-link', 'broken-anchor', 'invalid-path', 'anchor-validation-error'].forEach(type => {
      const issues = issuesByType[type] || [];
      if (issues.length === 0) return;

      const typeLabel = {
        'broken-link': 'Broken Internal Links',
        'broken-anchor': 'Broken Anchors',
        'invalid-path': 'Invalid Paths',
        'anchor-validation-error': 'Anchor Validation Errors',
      }[type];

      console.log(`\n📌 ${typeLabel} (${issues.length}):`);
      issues.forEach(issue => {
        console.log(`\n   File: ${issue.file}:${issue.line}`);
        console.log(`   Link: [${issue.text}](${issue.link})`);
        console.log(`   Error: ${issue.message}`);
        
        if (issue.availableHeadings && issue.availableHeadings.length > 0) {
          console.log(`   Available anchors: ${issue.availableHeadings.join(', ')}`);
        }
      });
    });
  }

  // Report external issues
  if (externalIssues.length > 0) {
    console.log('\n\n❌ EXTERNAL LINK ISSUES:\n');
    console.log(`🌐 Broken External Links (${externalIssues.length}):`);
    externalIssues.forEach(issue => {
      console.log(`\n   File: ${issue.file}:${issue.line}`);
      console.log(`   Link: [${issue.text}](${issue.link})`);
      console.log(`   Error: ${issue.error}`);
    });
  }

  // Summary
  console.log('\n' + '='.repeat(70));
  console.log('📊 SUMMARY\n');
  console.log(`Total files scanned: ${allFiles.length}`);
  console.log(`Internal link issues: ${internalIssues.length}`);
  if (config.checkExternal) {
    console.log(`External links checked: ${externalLinks.length}`);
    console.log(`External link issues: ${externalIssues.length}`);
  }
  console.log(`\nTotal issues: ${allIssues.length}`);
  console.log('='.repeat(70) + '\n');

  return {
    success: false,
    totalIssues: allIssues.length,
    internalIssues,
    externalIssues,
  };
}

// ============================================================================
// EXECUTION
// ============================================================================

(async () => {
  try {
    const result = await runLinkChecker();
    process.exit(result.success ? 0 : 1);
  } catch (err) {
    console.error(`Fatal error: ${err.message}`);
    process.exit(1);
  }
})();
